<a href="https://colab.research.google.com/github/JuanManuelLopezP/Anal-tica/blob/main/NaiveBayes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Modelo de preaprobación - Naive Bayes**
  Paso 0: Se cargan las librerias de trabajo

In [ ]:
import numpy as np #Libreria para funciones matemáticas
import pandas as pd #Libreria para hacer la comunicación con excel
from sklearn.naive_bayes import GaussianNB #Función Gaussiana
from sklearn.metrics import confusion_matrix #La métrica de medida

1. Se cargan los datos

In [ ]:
nxl= '/content/0. SolicitantesCrédito(USD)_NaiveBayes(L).xlsx'
XDB=pd.read_excel(nxl,sheet_name=2)  #Traer los datos
XDB.dropna() #Eliminar datos faltantes


#Seleccionamos las variables de interes
XDB=XDB[['Edad','Ingresos','Egresos','Cuota (COP)','PreApr']]
 #Mostrar la base de datos 'print(XDB)'

#Datos de entrada - np.array() convertir números
XD=np.array(XDB[['Edad','Ingresos','Egresos','Cuota (COP)']])
#Datos salida - Valor deseado - Que quiero que aprenda el modelo
yd=np.array(XDB['PreApr'])




#Para convertir la tabla en números
XDB=np.array(XDB)
print(XDB)


[[3.80000000e+01 1.35614400e+03 1.68562267e+03 1.83758903e+01
  1.00000000e+00]
 [5.10000000e+01 2.86016000e+02 3.64462000e+02 5.06249169e+00
  1.00000000e+00]
 [3.70000000e+01 5.17463250e+02 6.29208889e+02 1.02876881e+01
  0.00000000e+00]
 ...
 [3.80000000e+01 7.73019750e+02 6.72910667e+02 2.15087916e+01
  1.00000000e+00]
 [4.30000000e+01 6.35501750e+02 7.80691556e+02 1.10474685e+01
  1.00000000e+00]
 [3.70000000e+01 1.41926675e+03 1.16983844e+03 3.63561819e+01
  1.00000000e+00]]


2. Implementar modelo

In [ ]:
mnb=GaussianNB()   #Llamamos el modelo
mnb.fit(XD,yd)
ydp=mnb.predict(XD) #Muestra los datos de salida - pero que el obtuvo

#Parametros de modelo por categoría - PreaApr/PreNeg
u=mnb.theta_    #Se estiman las medias
sigma=np.sqrt(mnb.var_)   #Se estiman las varianzas

#las variaciones de las variables
LI=u-sigma
LS=u+sigma
print("La mdeia de las variables por categoría")
print(u)
print("El limite inferior de la variación de las variables por categoría es")
print(LI)
print("El limite superior de la variación de las variables por categoría es")
print(LS)


#Métrica de evaluación - Como se comportó el model identificando la PreApr-PreNeg
cm=confusion_matrix(yd,ydp)

array([[2563,  396],
       [1192, 1691]])

3. Se procede con la evaluación del modelo

In [ ]:
from re import S
VN=cm[0,0] #Verdaderos Negativos
FP=cm[0,1] #Falsos Positivos
FN=cm[1,0] #Falsos Negativos
VP=cm[1,1] #Verdadeos Positivos

Exactitud=(VP+VN)/len(XD)
print("La exactitud es:",Exactitud) #Comportamiento general

TE=(FN+FP)/len(XD)
print("El Tasa de Error es:",TE) #Porcentaje equivocaciones

Sensibilidad=VP/(VP+FP)
print("La sensibilidad es:",Sensibilidad) #Como se comportó frente a los positivos

Especificidad=VN/(VN+FN)
print("La especificidad es:",Especificidad) #Como se comportó frente a los negativos

Precisión=VP/(VP+FN)
print("La precisión es:",Precisión) #Como le pegó a los clasificados positivos

PredNeg=VN/(VN+FP)
print("La probabilidad de Negativo es:",PredNeg) #Como le pegó a los clasificados neagtivos

La exactitud es: 0.7281752824375214
El Tasa de Error es: 0.2718247175624786
La sensibilidad es: 0.8102539530426449
La especificidad es: 0.6825565912117177
La precisión es: 0.5865417967395075
La probabilidad de Negativo es: 0.8661710037174721


5. Que pasa con una sola persona

In [ ]:
Xun=np.array([[38,1356,1686,18]])  #Misma persona del excel
ydun=mnb.predict(Xun)
print(ydun)

[1]


Análisis de resultados:
De los resultados se puede observar que los preaprobados en general presentan edades, ingresos, egresos y cuotas mayores a los aprobados.Con respecto a las métricas se puede observar que el modelo funcionó mejor clasificando los positivos, como lo dice el indicador de sensibilidad el cual alcanzó el 81% de certeza. De manera general se puede observar que el modelo logró una exactitud que se ubicó por encima del 70% en promeido, y una tasa de error cercana al 30% en promedio. De manera general esta exactitud lo convierte en un buen modelo. Con respecto al solicitante de crédito (solicitante), este pertenece a la categoría de aprobado